core/domain/stats_jobs_continuous.py

```diff
@@ -411,27 +412,72 @@ def entity_classes_to_map_over(cls):
 
     @staticmethod
     def map(item):
+        # pylint: disable=line-too-long
         if InteractionAnswerSummariesMRJobManager._entity_created_before_job_queued(
                 item):
+            state_answers_dict = {
+                'exploration_id': item.exploration_id,
+                'exploration_version': item.exploration_version,
+                'state_name': item.state_name,
+                'interaction_id': item.interaction_id,
+                'submitted_answer_list': copy.deepcopy(
+                    item.submitted_answer_list)
+            }
 
-            # All visualizations desired for the interaction.
-            visualizations = interaction_registry.Registry.get_interaction_by_id(
-                item.interaction_id).answer_visualizations
-
-            # Get all desired calculations for the current state interaction id.
-            calc_ids = list(set(viz.calculation_id for viz in visualizations))
-            calculations = [
-                calculation_registry.Registry.get_calculation_by_id(calc_id)
-                for calc_id in calc_ids]
+            # Output answers submitted to the exploration for this exp version.
+            yield ('%s:%s:%s' % (
+                item.exploration_id, item.exploration_version,
+                item.state_name), state_answers_dict)
 
-            # Perform each calculation, and store the output.
-            for calc in calculations:
-                calc_output = calc.calculate_from_state_answers_entity(item)
-                calc_output.save()
+            # Output the same set of answers independent of the version. This
+            # allows the reduce step to aggregate answers across all
+            # exploration versions.
+            state_answers_dict['exploration_version'] = VERSION_ALL
+            yield ('%s:%s:%s' % (
+                item.exploration_id, VERSION_ALL,
+                item.state_name), state_answers_dict)
 
     @staticmethod
-    def reduce(id, state_answers_model):
-        pass
+    def reduce(key, stringified_submitted_answer_list):
+        exploration_id, exploration_version, state_name = key.split(':')
+        interaction_id = ast.literal_eval(
+            stringified_submitted_answer_list[0])['interaction_id']
+
+        # Collapse the list of answers into a single answer dict. This
+        # aggregates across multiple answers if the key ends with VERSION_ALL.
+        combined_state_answers = {
+            'exploration_id': exploration_id,
+            'exploration_version': exploration_version,
+            'state_name': state_name,
+            'interaction_id': interaction_id,
+            'submitted_answer_list': list(itertools.chain.from_iterable(
+                ast.literal_eval(state_answers)['submitted_answer_list']
+                for state_answers in stringified_submitted_answer_list))
+        }
+
+        # NOTE TO DEVELOPERS: 'submitted_answer_list' above is being converted
```

Comment:
Yeah, I kind of wonder how large this could get. Any guarantee of an upper bound? Could this be causing your job to breach the memory limit?